### PR TITLE
Implements --exclude-file

### PR DESCRIPTION
**Pull request by dblock**
Implemented https://github.com/lycheeverse/lychee/issues/302.

This is my first rust code, so talk to me like a 2yo :)

### CODE REVIEW

File: lychee-bin/src/main.rs [hunk 1]
```diff
@@ -113,6 +115,18 @@ fn run_main() -> Result<i32> {
...
         opts.config.merge(c)
     }
+
+    // Load excludes from file
+    for path in &opts.config.exclude_file {
+        let file = File::open(path).expect("No such file");
+        opts.config.exclude.append(
+            &mut io::BufReader::new(file)
+                .lines()
+                .map(|l| l.expect("Could not read line"))
```
Comment (lebensterben) [inline lychee-bin/src/main.rs:125]: there's no need to manually unwrap a Result in Iterator.
If you've a Itertor<Item = Result<T, E>>, you can collect it directly to a Result<Vec<T>, E>
- If there's any Err, it returns an Err
- If all elements arr Ok, it returns Ok(Vec<T>)

**Comment by lebensterben** [comment]
I think your original proposal of `.lycheeignore` is great.
We should make it default.

And this also allows the possibility for per-directory exclusion list.

**Comment by dblock** [comment]
> I think your original proposal of `.lycheeignore` is great.
> We should make it default.
> 
> And this also allows the possibility for per-directory exclusion list.

It's a new construct for Lychee, so I think we should think about it more. The thing with things like `.gitignore` is that they git walks the parent path to find the closest .gitignore until it reaches the git root. This won't be possible here, so maybe it's too much magic? 

One way to implement that would be to default --exclude-file to `.lycheeexclude` if that exists. Can be built on top of this PR.

**Comment by lebensterben** [comment]
>  The thing with things like .gitignore is that git walks the parent path to find the closest .gitignore until it reaches the git root. 

lychee supports 'glob'.
so you can do something like `lychee ~/notes/**/*.md`.
when expanding the glob pattern it's already walking the directories.

File: lychee-bin/src/main.rs [hunk 2]
```diff
@@ -113,6 +115,18 @@ fn run_main() -> Result<i32> {
     if let Some(c) = Config::load_from_file(&opts.config_file)? {
         opts.config.merge(c)
     }
+
+    // Load excludes from file
+    for path in &opts.config.exclude_file {
+        let file = File::open(path).expect("No such file");
```
Comment (mre) [inline lychee-bin/src/main.rs:121]: Instead of `.expect (...)` can you use a `?` here? Makes it a little cleaner.

**Review by mre — CHANGES_REQUESTED** [review]

**Comment by mre** [comment]
Thanks for the PR. 
I also like the idea of a `.lycheeignore`. In contrast to the `.gitignore`, a `.dockerignore` will only be parsed from the build context (usually the current directory), so I would not be too concerned about the differences between the implementations as long as we document the behavior and add tests. 

As you both mentioned it could be a nice addition in the future.

File: lychee-bin/src/main.rs (hunk 1 above)
Comment (mre) [reply lychee-bin/src/main.rs]: @dblock I resolved this by pushing to your branch if you want to have a look at the diff.

File: lychee-bin/src/main.rs (hunk 2 above)
Comment (mre) [reply lychee-bin/src/main.rs]: Resolved by pushing my changes to the branch.

File: lychee-bin/tests/cli.rs [hunk 3]
```diff
@@ -364,10 +364,76 @@ mod cli {
...
+        let expected = r#"{"total":10,"successful":10,"failures":0,"timeouts":0,"redirects":0,"excludes":0,"errors":0,"fail_map":{}}"#;
         let output = fs::read_to_string(&outfile)?;
         assert_eq!(output.split_whitespace().collect::<String>(), expected);
         fs::remove_file(outfile)?;
         Ok(())
     }
+
+    /// Test excludes
+    #[test]
+    fn test_exclude_wildcard() -> Result<()> {
```
Comment (mre) [inline lychee-bin/tests/cli.rs:376]: Thanks for adding these additional exclude tests and increasing the test coverage. ❤️

**Comment by mre** [comment]
I'll merge this now. Expect a new release soon. Created a ticket for discussing the future addition of `.lycheeignore` at https://github.com/lycheeverse/lychee/issues/308. 
Thanks for first (but hopefully not last) your contribution @dblock. ⭐ And welcome to the Rust community! 🦀

**Pull request merged**